src/main/java/org/ibp/api/java/impl/middleware/common/validator/GermplasmUpdateValidator.java

```diff
@@ -0,0 +1,161 @@
+import org.apache.commons.lang3.StringUtils;
+import org.generationcp.commons.util.DateUtil;
+import org.generationcp.middleware.api.attribute.AttributeDTO;
+import org.generationcp.middleware.api.breedingmethod.BreedingMethodService;
+import org.generationcp.middleware.api.germplasm.GermplasmNameTypeDTO;
+import org.generationcp.middleware.domain.germplasm.GermplasmUpdateDTO;
+import org.generationcp.middleware.manager.api.GermplasmDataManager;
+import org.generationcp.middleware.manager.api.LocationDataManager;
+import org.generationcp.middleware.pojos.Germplasm;
+import org.generationcp.middleware.pojos.Location;
+import org.ibp.api.java.germplasm.GermplasmService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import org.springframework.validation.BindingResult;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Component
+public class GermplasmUpdateValidator {
+
+	@Autowired
+	private GermplasmService germplasmService;
+
+	@Autowired
+	private org.generationcp.middleware.api.germplasm.GermplasmService germplasmMiddlewareService;
+
+	@Autowired
+	private GermplasmDataManager germplasmDataManager;
+
+	@Autowired
+	private LocationDataManager locationDataManager;
+
+	@Autowired
+	private BreedingMethodService breedingMethodService;
+
+	public void validateEmptyList(final BindingResult errors, final List<GermplasmUpdateDTO> germplasmUpdateDTOList) {
+		if (germplasmUpdateDTOList == null || germplasmUpdateDTOList.isEmpty()) {
+			errors.reject("germplasm.update.empty.list", "");
+		}
+	}
+
+	public void validateCodes(final BindingResult errors, final List<GermplasmUpdateDTO> germplasmUpdateDTOList) {
+
+		final Set<String> nameCodes = new HashSet<>();
+		germplasmUpdateDTOList
+			.forEach(g -> nameCodes.addAll(g.getNames().keySet().stream().map(n -> n.toUpperCase()).collect(Collectors.toList())));
+		nameCodes.addAll(
+			germplasmUpdateDTOList.stream()
+				.map(o -> StringUtils.isNotEmpty(o.getPreferredName()) ? o.getPreferredName().toUpperCase() : null)
+				.filter(Objects::nonNull).collect(
+				Collectors.toSet()));
+		final List<String> existingNamesCode =
+			this.germplasmService.filterGermplasmNameTypes(nameCodes).stream().map(GermplasmNameTypeDTO::getCode).collect(
+				Collectors.toList());
+
+		final Set<String> attributesCode = new HashSet<>();
+		germplasmUpdateDTOList.stream().filter(germ -> germ.getAttributes() != null).collect(Collectors.toList())
+			.forEach(
+				g -> attributesCode.addAll(g.getAttributes().keySet().stream().map(n -> n.toUpperCase()).collect(Collectors.toList())));
+		final List<String> existingAttributesCode =
+			this.germplasmService.filterGermplasmAttributes(attributesCode).stream().map(AttributeDTO::getCode).collect(
+				Collectors.toList());
+
+		nameCodes.removeAll(existingNamesCode);
```

Comment:
Is it sufficient to do `nameCodes.equals(existingNamesCodes)` and `attributesCode .equals(existingAttributesCodes) `